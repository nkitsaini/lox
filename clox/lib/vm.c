#include "vm.h"
#include "common.h"
#include "compiler.h"
#include "debug.h"
#include "memory.h"
#include "string.h"
#include <stdarg.h>
VM vm;

static void resetStack()
{
	vm.stackTop = vm.stack;
}

static void runtimeError(const char *format, ...)
{
	// TODO: I don't understand this. I wrote it just by looking at book
	va_list args;
	va_start(args, format);
	vfprintf(stderr, format, args);
	va_end(args);
	fputs("\n", stderr);
	size_t instruction = vm.ip - vm.chunk->code - 1;
	int line = vm.chunk->lines[instruction];
	fprintf(stderr, "[line %d] in script\n", line);
	resetStack();
}

static Value peek(int distance)
{
	return vm.stackTop[-1 - distance];
}

static bool isFalsey(Value value)
{
	return IS_NIL(value) || (IS_BOOL(value) && !AS_BOOL(value));
}
static void concatenate()
{
	ObjString *b = AS_STRING(pop());
	ObjString *a = AS_STRING(pop());
	int length = a->length + b->length;
	char *chars = ALLOCATE(char, length + 1);
	memcpy(chars, a->chars, a->length);
	memcpy(chars + a->length, b->chars, b->length);
	chars[length] = '\0';
	ObjString *result = takeString(chars, length);
	push(OBJ_VAL(result));
}

void initVM()
{
	vm.stack = NULL;
	vm.stack_length = 0;
	vm.objects = NULL;
	resetStack();
}

void freeVM()
{
	// INVESTIGATE: Chapter 19 does not mention other code, how did I get it?
	FREE_ARRAY(Value, vm.stack, vm.stack_length);
	vm.stack_length = 0;
	vm.stack = NULL;
	vm.stackTop = NULL;
	freeObjects();
}

void push(Value value)
{
	if (vm.stack_length == 0 || vm.stackTop == (vm.stack + vm.stack_length))
	{
		int new_size = GROW_CAPACITY(vm.stack_length);
		vm.stack = GROW_ARRAY(Value, vm.stack, vm.stack_length, new_size);
		vm.stackTop = vm.stack + vm.stack_length;
		vm.stack_length = new_size;
	}
	*vm.stackTop = value;
	vm.stackTop++;
}

Value pop()
{
	vm.stackTop--;
	return *vm.stackTop;
}

static InterpretResult run()
{
#define READ_BYTE() (*vm.ip++)
#define READ_CONSTANT() (vm.chunk->constants.values[READ_BYTE()])
// we pop `b` first because it'll be inserted last.
// In expression `3 + 1`, we'll first insert `3` and then `1` to stack
// So we need to assign `1` (top most element) to b and `3` to a
#define BINARY_OP(valueType, op)                        \
	do                                                  \
	{                                                   \
		if (!IS_NUMBER(peek(0)) || !IS_NUMBER(peek(1))) \
		{                                               \
			runtimeError("Operands must be numbers;");  \
			return INTERPRET_RUNTIME_ERROR;             \
		}                                               \
		double b = AS_NUMBER(pop());                    \
		double a = AS_NUMBER(pop());                    \
		push(valueType(a op b));                        \
	} while (false)

	for (;;)
	{

#ifdef DEBUG_TRACE_EXECUTION
		printf("        ");
		for (Value *slot = vm.stack; slot < vm.stackTop; slot++)
		{
			printf("[ ");
			printValue(*slot);
			printf(" ]");
		}
		printf("\n");
		disassembleInstruction(vm.chunk, (int)(vm.ip - vm.chunk->code));
#endif
		uint8_t instruction;
		switch (instruction = READ_BYTE())
		{
		case OP_CONSTANT:
		{
			Value constant = READ_CONSTANT();
			printValue(constant);
			push(constant);
			printf("\n");
			break;
		}
		case OP_NEGATE:
		{
			// Little bit faster then push(-pop());
			// Used hyperfine (500_000 negations create chunk + disassemble + execute chunk)
			//
			// Using inplace modification:
			//     Time (mean ± σ):     534.5 ms ±   5.4 ms    [User: 508.4 ms, System: 36.3 ms]
			//     Range (min … max):   525.0 ms … 543.8 ms    10 runs
			// Using Pop:
			//     Time (mean ± σ):     549.7 ms ±   7.9 ms    [User: 526.0 ms, System: 34.5 ms]
			//     Range (min … max):   542.2 ms … 567.1 ms    10 runs

			if (!IS_NUMBER(peek(0)))
			{
				runtimeError("Operand must be a number");
				return INTERPRET_RUNTIME_ERROR;
			}
			push(NUMBER_VAL(-AS_NUMBER(pop())));
			break;
		}
		case OP_NOT:
		{
			push(BOOL_VAL(isFalsey(pop())));
			break;
		}
		case OP_NIL:
			push(NIL_VAL);
			break;
		case OP_TRUE:
			push(BOOL_VAL(true));
			break;
		case OP_FALSE:
			push(BOOL_VAL(false));
			break;
		case OP_EQUAL:
		{
			Value b = pop();
			Value a = pop();
			push(BOOL_VAL(valuesEqual(a, b)));
			break;
		}
		case OP_GREATER:
			BINARY_OP(BOOL_VAL, >);
			break;
		case OP_LESS:
			BINARY_OP(BOOL_VAL, <);
			break;
		case OP_ADD:
		{
			if (IS_STRING(peek(0)) && IS_STRING(peek(1)))
			{
				concatenate();
			}
			else if (IS_NUMBER(peek(0)) && IS_NUMBER(peek(1)))
			{
				double b = AS_NUMBER(pop());
				double a = AS_NUMBER(pop());
				push(NUMBER_VAL(a + b));
			}
			else
			{
				runtimeError("Operands must be two numbers or two strings.");
				return INTERPRET_RUNTIME_ERROR;
			}
			break;
		}
		case OP_SUBTRACT:
			BINARY_OP(NUMBER_VAL, -);
			break;
		case OP_MULTIPLY:
			BINARY_OP(NUMBER_VAL, *);
			break;
		case OP_DIVIDE:
			BINARY_OP(NUMBER_VAL, /);
			break;
		case OP_RETURN:
		{
			printValue(pop());
			return INTERPRET_OK;
		}
		}
	}

#undef READ_BYTE
#undef READ_CONSTANT
#undef BINARY_OP
}

InterpretResult interpret(char *source)
{
	Chunk chunk;
	initChunk(&chunk);

	if (!compile(source, &chunk))
	{
		freeChunk(&chunk);
		return INTERPRET_COMPILE_ERROR;
	}
	vm.chunk = &chunk;
	vm.ip = vm.chunk->code;
	InterpretResult result = run();
	freeChunk(&chunk);
	return result;
}
