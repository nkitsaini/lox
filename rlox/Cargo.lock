# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "anyhow"
version = "1.0.71"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9c7d0618f0e0b7e8ff11427422b64564d5fb0be1940354bfe2e0529b18a9d9b8"

[[package]]
name = "enum-as-inner"
version = "0.6.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5ffccbb6966c05b32ef8fbac435df276c4ae4d3dc55a8cd0eb9745e6c12f546a"
dependencies = [
 "heck",
 "proc-macro2",
 "quote",
 "syn 2.0.18",
]

[[package]]
name = "enum-kinds"
version = "0.5.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4e40a16955681d469ab3da85aaa6b42ff656b3c67b52e1d8d3dd36afe97fd462"
dependencies = [
 "proc-macro2",
 "quote",
 "syn 1.0.109",
]

[[package]]
name = "heck"
version = "0.4.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "95505c38b4572b2d910cecb0281560f54b440a19336cbbcb27bf6ce6adc6f5a8"

[[package]]
name = "proc-macro2"
version = "1.0.60"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "dec2b086b7a862cf4de201096214fa870344cf922b2b30c167badb3af3195406"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.28"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1b9ab9c7eadfd8df19006f1cf1a4aed13540ed5cbc047010ece5826e10825488"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "rlox"
version = "0.1.0"
dependencies = [
 "anyhow",
 "enum-as-inner",
 "enum-kinds",
 "smallvec",
]

[[package]]
name = "smallvec"
version = "1.10.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a507befe795404456341dfab10cef66ead4c041f62b8b11bbb92bffe5d0953e0"

[[package]]
name = "syn"
version = "1.0.109"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "72b64191b275b66ffe2469e8af2c1cfe3bafa67b529ead792a6d0160888b4237"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "syn"
version = "2.0.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "32d41677bcbe24c20c52e7c70b0d8db04134c5d1066bf98662e2871ad200ea3e"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "unicode-ident"
version = "1.0.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b15811caf2415fb889178633e7724bad2509101cde276048e013b9def5e51fa0"
