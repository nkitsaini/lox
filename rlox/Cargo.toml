[package]
name = "rlox"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
# Defines a feature named `webp` that does not enable any other features.
trace_execution = []
default = ["trace_execution"]

[dependencies]
anyhow = "1.0.71"
enum-as-inner = "0.6.0"
smallvec = "1.10.0"
