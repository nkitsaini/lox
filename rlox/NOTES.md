


# Data Location

At parser/compilation all the string refer to the source code representation.
When they move to VM they're owned by us and should be present in Interned Strings.



