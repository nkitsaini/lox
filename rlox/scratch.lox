var a = 3;
a  = a + 2;

var b = 9;

{
	var b = b;
}

print a;
print b;


