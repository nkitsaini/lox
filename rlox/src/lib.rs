pub mod chunk;
pub mod compiler;
pub mod debug;
pub mod hashtable;
pub mod prelude;
pub mod scanner;
pub mod value;
pub mod vm;
