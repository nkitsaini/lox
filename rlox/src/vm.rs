// use crate::{compiler::compile, prelude::*};
use crate::{compiler::Compiler, hashtable::HashTable, prelude::*, value::LoxObject};
use smallvec;

const STACK_MAX: usize = 256;

pub struct VM<'a> {
    chunk: Chunk<'a>,

    // Huh, the book says looking by index is slower them
    // looking by index. Why could that be? Due to additions?
    ip: usize,

    stack: smallvec::SmallVec<[Value<'a>; STACK_MAX]>,

    strings: HashTable<'a>,
}

fn is_falsey(value: Value) -> bool {
    match value {
        Value::Nil => true,
        Value::Bool(x) => !x,
        _ => false,
    }
}

// fn values_equal(value1: Value, value2: Value) -> bool {
//     if value1.get_type() != value2.get_type() {
//         return false;
//     }

// }
macro_rules! binary_op {
    ($vm:ident, $result:path, $op:tt) => {{
        let (a, b) = match ($vm.peek(0), $vm.peek(1)) {
            (Value::Number(a), Value::Number(b)) => (a, b),
            _ => {
                $vm.runtime_error("Operands must be numbers.");
                return Err(InterpreterError::RuntimeError);
            }
        };
        $vm.stack.pop();
        $vm.stack.pop();

        // $vm.stack.push(Value::Number(b $op a));

        $vm.stack.push($result(b $op a));
    }};
}

impl<'a> VM<'a> {
    pub fn new(chunk: Chunk<'a>) -> Self {
        Self {
            chunk,
            ip: 0,
            stack: smallvec::SmallVec::new(),
            strings: HashTable::new(),
        }
    }

    pub fn new_with_strings(chunk: Chunk<'a>, strings: HashTable<'a>) -> Self {
        Self {
            chunk,
            ip: 0,
            stack: smallvec::SmallVec::new(),
            strings,
        }
    }

    pub fn interpret(source: &'a str) -> InterpreterResult {
        let strings = HashTable::new();
        let (chunk, strings) = match Compiler::compile(source, strings) {
            Some(x) => x,
            None => return Err(InterpreterError::CompileError),
        };

        let mut vm = Self::new_with_strings(chunk, strings);
        let result = vm.run();
        return result;
    }

    fn peek(&self, distance: usize) -> Value {
        let v = (-1 - distance as i32) as usize;
        return self.stack.get(v % self.stack.len()).cloned().unwrap();
    }

    fn run(&mut self) -> InterpreterResult {
        loop {
            let instruction = self.chunk.code[self.ip].0;

            // Debug stuff
            #[cfg(feature = "trace_execution")]
            {
                print!("        stack: ");
                for value in self.stack.iter() {
                    print!("[ ");
                    value.print();
                    print!(" ]");
                }
                println!();
                instruction.show_disassemble(&self.chunk);
            }

            use OpCode::*;

            // Execution
            self.ip += 1;
            match instruction {
                Return => {
                    self.stack.pop().unwrap().print();
                    println!();
                    return Ok(());
                }
                Constant { location } => {
                    let constant = self.chunk.constants[location as usize].clone();
                    self.stack.push(constant);
                }
                Negate => {
                    let val = match self.peek(0) {
                        Value::Number(x) => x,
                        _ => {
                            self.runtime_error("Operand must be a number.");
                            return Err(InterpreterError::RuntimeError);
                        }
                    };
                    self.stack.push(Value::Number(val));
                }

                Nil => self.stack.push(Value::Nil),
                True => self.stack.push(Value::Bool(true)),
                False => self.stack.push(Value::Bool(false)),

                Equal => {
                    let a = self.stack.pop().unwrap();
                    let b = self.stack.pop().unwrap();
                    self.stack.push(Value::Bool(a == b));
                }
                Greater => binary_op!(self, Value::Bool, >),
                Less => binary_op!(self, Value::Bool, <),

                Not => {
                    let val = self.stack.pop().unwrap();
                    self.stack.push(Value::Bool(is_falsey(val)));
                }
                Add => match (self.peek(0), self.peek(1)) {
                    (Value::Number(x), Value::Number(y)) => {
                        self.stack.pop();
                        self.stack.pop();
                        self.stack.push(Value::Number(y + x));
                    }
                    (Value::Object(a), Value::Object(b)) => match (a.as_ref(), b.as_ref()) {
                        (
                            LoxObject::String { value: _, hash: _ },
                            LoxObject::String { value: _, hash: _ },
                        ) => {
                            self.concatenate();
                        }
                        _ => {
                            self.runtime_error("Operands must be two numbers or two strings.");
                            return Err(InterpreterError::RuntimeError);
                        }
                    },
                    _ => {
                        self.runtime_error("Operands must be two numbers or two strings.");
                        return Err(InterpreterError::RuntimeError);
                    }
                },
                // Add => binary_op!(self, Value::Number, +),
                Multiply => binary_op!(self, Value::Number, *),
                Subtract => binary_op!(self, Value::Number, -),
                Divide => binary_op!(self, Value::Number, /),
            }
        }
    }

    fn runtime_error(&mut self, msg: &str) {
        eprintln!("{}", msg);

        let instruction = self.ip - 1;
        let line = self.chunk.code[instruction].1;
        eprintln!("[line {}] in script\n", line);
        self.stack.clear();
    }

    fn concatenate(&mut self) {
        let a = self.stack.pop().unwrap();
        let b = self.stack.pop().unwrap();
        let a = a.as_object().unwrap().as_string().unwrap();
        let b = b.as_object().unwrap().as_string().unwrap();
        let result = b.0.to_string() + &a.0;

        let res = self.allocate_string(result);
        self.stack.push(Value::Object(res));
    }

    fn allocate_string(&mut self, val: String) -> Rc<LoxObject<'a>> {
        let lox_str = LoxObject::new_string(val);
        let entry = self.strings.find_string(&lox_str).clone();
        if let Some(x) = entry {
            return x;
        }
        let str = Rc::new(lox_str);
        self.strings.set(str.clone(), Value::Nil);
        return str;
    }
}

#[derive(Debug)]
pub enum InterpreterError {
    CompileError,
    RuntimeError,
}

pub type InterpreterResult = Result<(), InterpreterError>;
