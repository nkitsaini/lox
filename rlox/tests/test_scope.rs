use std::io::Cursor;

use pretty_assertions::assert_eq;
use rlox::vm::{InterpreterError, VM};

#[test]
fn test_scope() {
    const SCRIPT: &'static str = "
var a = 3;
a  = a + 2;

var b = 9;

{
	var b = 7;
	b = b + a;
	print b;
}

print a;
print b;
";
    let mut stdout = Cursor::new(Vec::new());
    let mut stderr = Cursor::new(Vec::new());
    let mut vm = VM::empty_new(&mut stdout, &mut stderr);
    let res = vm.interpret(SCRIPT);

    drop(vm);

    assert_eq!(res, Ok(()));
    assert!(stderr.into_inner().is_empty());
    let stdout = stdout.into_inner();
    let out = String::from_utf8(stdout).unwrap();
    insta::assert_snapshot!(out);
}

#[test]
fn test_scope_error() {
    const SCRIPT: &'static str = "
var a = 3;
a  = a + 2;

var b = 9;

{
	var b = b; // this should show error
}

print a;
print b;
";
    let mut stdout = Cursor::new(Vec::new());
    let mut stderr = Cursor::new(Vec::new());
    let mut vm = VM::empty_new(&mut stdout, &mut stderr);
    let res = vm.interpret(SCRIPT);

    drop(vm);

    assert_eq!(res, Err(InterpreterError::CompileError));
    insta::assert_snapshot!(String::from_utf8(stderr.into_inner()).unwrap());
    insta::assert_snapshot!(String::from_utf8(stdout.into_inner()).unwrap());
}
