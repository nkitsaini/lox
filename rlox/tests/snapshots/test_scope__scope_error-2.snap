---
source: tests/test_scope.rs
expression: "String::from_utf8(stdout.into_inner()).unwrap()"
---

